source/module_hamilt_lcao/hamilt_lcaodft/operator_lcao/ekinetic_new.cpp

```diff
@@ -0,0 +1,264 @@
+    const int* iw2l2 = atom2.iw2l;
+    const int* iw2n2 = atom2.iw2n;
+    const int* iw2m2 = atom2.iw2m;
+    // ---------------------------------------------
+    // get tau1 (in cell <0,0,0>) and tau2 (in cell R)
+    // in principle, only dtau is needed in this function
+    // snap_psipsi should be refactored to use dtau directly
+    // ---------------------------------------------
+    const ModuleBase::Vector3<double>& tau1 = this->ucell->get_tau(iat1);
+    const ModuleBase::Vector3<double> tau2 = tau1 + dtau;
+    // ---------------------------------------------
+    // calculate the Ekinetic matrix for each pair of orbitals
+    // ---------------------------------------------
+    double olm[3] = {0, 0, 0};
+    auto row_indexes = paraV->get_indexes_row(iat1);
+    auto col_indexes = paraV->get_indexes_col(iat2);
+    const int step_trace = col_indexes.size() + 1;
+    for (int iw1l = 0; iw1l < row_indexes.size(); iw1l += npol)
+    {
+        const int iw1 = row_indexes[iw1l] / npol;
+        const int L1 = iw2l1[iw1];
+        const int N1 = iw2n1[iw1];
+        const int m1 = iw2m1[iw1];
+        for (int iw2l = 0; iw2l < col_indexes.size(); iw2l += npol)
+        {
+            const int iw2 = col_indexes[iw2l] / npol;
+            const int L2 = iw2l2[iw2];
+            const int N2 = iw2n2[iw2];
+            const int m2 = iw2m2[iw2];
+            uot.snap_psipsi(orb, // orbitals
+                            olm,
+                            0,
+                            'T', // olm, job of derivation, dtype of Operator
+                            tau1,
+                            T1,
+                            L1,
+                            m1,
+                            N1, // info of atom1
+                            tau2,
+                            T2,
+                            L2,
+                            m2,
+                            N2 // info of atom2
+            );
+            for (int ipol = 0; ipol < npol; ipol++)
+            {
+                data_pointer[ipol * step_trace] += olm[0];
+            }
+            data_pointer += npol;
+        }
+        data_pointer += (npol - 1) * col_indexes.size();
+    }
+}
+
+// set_HR_fixed()
+template <typename TK, typename TR>
+void hamilt::EkineticNew<hamilt::OperatorLCAO<TK>, TR>::set_HR_fixed(void* HR_fixed_in)
+{
+    this->HR_fixed = static_cast<hamilt::HContainer<TR>*>(HR_fixed_in);
+    this->allocated = false;
+}
+
+// contributeHR()
+template <typename TK, typename TR>
+void hamilt::EkineticNew<hamilt::OperatorLCAO<TK>, TR>::contributeHR()
+{
+    if (!this->HR_fixed_done)
+    {
+        // if this Operator is the first node of the sub_chain, then HR_fixed is nullptr
+        if (this->HR_fixed == nullptr)
+        {
+            this->HR_fixed = new hamilt::HContainer<TR>(*this->HR);
+            this->allocated = true;
+        }
+        else
+        {
+            this->HR_fixed->synchronize(*this->HR);
```

Comment:
I suggest change the function name to "padding(*this->HR, 0)" or "padding_like(*this->HR, 0)" or something for improving the readability.

Reply:
Thanks, good suggestion!